src/coreclr/jit/emit.cpp

```diff
@@ -4102,6 +4199,382 @@ void emitter::emitDispCommentForHandle(size_t handle, GenTreeFlags flag)
+// emitRemoveJumpToNextInst:  Checks all jumps in the jump list to see if they are
+//   unconditional jumps generated at the end of BB_ALWAYS basic blocks and whether
+//   they are now a jump to the next instruction, this can happen when the following
+//   basic block emitted no instructions
+//
+// Returns:
+//   true if any jumps have been removed, otherwise returns false
+//
+// Assumptions:
+//    the jump list must be ordered by increasing igNum+insNo
+//
+bool emitter::emitRemoveJumpToNextInst()
+{
+#ifdef TARGET_XARCH
+#ifdef DEBUG
+    if (emitComp->verbose)
+    {
+        printf("*************** In emitRemoveJumpToNextInst()\n");
+    }
+#endif
+    if (emitJumpAlwaysList == nullptr)
+    {
+#ifdef DEBUG
+        if (EMITVERBOSE)
+        {
+            printf("emitRemoveJumpToNextInst emitJumpAlwaysList is empty, removed 0 jumps and 0 bytes\n");
+        }
+#endif
+        return false;
+    }
+#ifdef DEBUG
+    if (EMIT_INSTLIST_VERBOSE)
+    {
+        printf("\nInstruction group list before unconditional jump to next instruction removal:\n\n");
+        emitDispIGlist(true);
+    }
+    if (EMITVERBOSE)
+    {
+        printf("jump always list %u\n", emitJumpAlwaysList);
+        emitDispJumpList(emitJumpAlwaysList);
+    }
+#endif
+
+    int            removedCount           = 0;
+    UNATIVE_OFFSET removedSize            = 0;
+    instrDescJmp*  previousJmp            = nullptr;
+    insGroup*      previousTruncatedGroup = nullptr;
+    instrDescJmp*  lastInserted           = nullptr;
+#if DEBUG
+    UNATIVE_OFFSET previousJumpIgNo  = (UNATIVE_OFFSET)-1;
+    unsigned int   previousJumpInsNo = -1;
+#endif
+    assert(!emitJumpAlwaysListCompleted);
+    emitJumpAlwaysListCompleted = true;
+
+    while (emitJumpAlwaysList)
+    {
+        instrDescJmp* jmp = emitJumpAlwaysList;
+
+        emitJumpAlwaysList = jmp->idjNext;
+        jmp->idjNext       = nullptr;
+
+        // if the jump is unconditional then it is a candidate
+        if (jmp->idInsFmt() == IF_LABEL && emitIsUncondJump(jmp) && !jmp->idjKeepLong)
+        {
+#if DEBUG
+            assert(jmp->idjIG->igNum > previousJumpIgNo || previousJumpIgNo == (UNATIVE_OFFSET)-1 ||
+                   (jmp->idjIG->igNum == previousJumpIgNo && jmp->idDebugOnlyInfo()->idNum > previousJumpInsNo));
+            previousJumpIgNo  = jmp->idjIG->igNum;
+            previousJumpInsNo = jmp->idDebugOnlyInfo()->idNum;
+#endif
+
+            // target group is not bound yet so use the cookie to fetch it
+            insGroup* targetGroup = (insGroup*)emitCodeGetCookie(jmp->idAddr()->iiaBBlabel);
+
+            if (targetGroup != nullptr && (jmp->idjIG->igFlags & IGF_HAS_ALIGN) == 0 &&
+                jmp->idjIG->igNext == targetGroup)
+            {
+                insGroup* group            = jmp->idjIG;
+                unsigned  instructionCount = group->igInsCnt;
+
+                if (instructionCount > 0)
```

Comment:
I am not sure if I follow the reasoning here. `group` contains the `jmp` instruction, if not anything else, so `instructionCount` will always be `> 0`, right? Or am i missing something?

Reply:
I assume the `instructionCount ` will be removed once you use "zero the instruction size" approach.

Reply:
The check should have been >1, i've corrected that. I think the instruction count needs to remain because I want to make sure that the jump really is the last instruction in the group. Doing that check has already caught one codegen issue. If it causes a throughput issue it could be made debug only.

Reply:
> I want to make sure that the jump really is the last instruction in the group

Are you saying that you don't want `jmp` to be the only instruction in the group? I am not sure if I follow how does that check helps?

Reply:
Two cases. If jmp is the only instruction in the group I'm currently skipping it because I haven't seen that case and I'm not sure what happens if I remove the only instruction in a group. If the group has multiple instructions I want to make sure that the jmp I'm trying to remove is the last one.

Reply:
You can have this in debug-only check.

```c++
#ifdef DEBUG

instrDesc* instructionDescriptor = nullptr;
while (instructionCount > 0)
{
    instructionDescriptor = (instrDesc*)dataPtr;
    dataPtr += emitSizeOfInsDsc(instructionDescriptor);
    instructionCount -= 1;
};

assert(jmp == instructionDescriptor);
#endif
```

> Doing that check has already caught one codegen issue

Was that related to dead code in IG or was it something else?

Reply:
It was the dead code. I'm going to leave this in for now and come back to it later because i want to see if i can find a case of a single jmp containing IG which is going to take some time to do, there are a lot of diffs.

Reply:
I put an assertion in to fire on ig's with a single jmp always and there are hits in the tests. I'm going to isolate one and see what's going on. Then I'll be able to experiment with removal. This is WIP.